### PR TITLE
feat(executor): stream command stdout via chunk callback

### PR DESCRIPTION
**Pull request by jisung-02**
## Summary

- Stream system shell command stdout/stderr to alpacon-server in real time via a new `/api/events/commands/{id}/chunk/` endpoint, replacing the previous fin-only post-completion model.
- A `chunkWriter` emits chunks on newline boundaries or when the in-memory buffer crosses 4 KB, whichever comes first; partial lines are carried over between writes and flushed at completion.
- The `CommandRunner` owns the monotonic `seq` counter so multiple `chunkWriter` instances spawned across shell operators (`&&`, `||`, `;`) in `executeWithOperators` share one series per `command_id`, preventing the unique `(command, seq)` collision that the initial implementation had.
- Fin POST priority is bumped from 10 to 11 so pending chunk POSTs drain ahead of fin. The server marks the command handled on fin and drops late chunks, so ordering is enforced on the alpamon side.

## Changes

- `pkg/executor/executor.go`: new `chunkWriter` (sync `io.Writer`), `ChunkCallback` on `CommandOptions`, `ExecWithStreamingHook` entry point. `runCommand` routes through `chunkWriter` when streaming is requested and falls back to the existing `CombinedOutput` / `Start+Wait` paths otherwise.
- `pkg/executor/handlers/common`: `CommandExecutor.ExecWithStreamingHook` interface, `ChunkCallback` field on `CommandArgs`, mock implementation.
- `pkg/executor/handlers/shell/shell.go`: forwards `ChunkCallback` through `executeCommand` and `executeWithOperators`, choosing the streaming or non-streaming executor entry point per call.
- `pkg/runner/client.go`: new `eventCommandChunkURL` constant.
- `pkg/runner/command.go`: builds the per-command chunk callback closure (owns `seq`), posts each chunk to the chunk URL, and lowers fin priority to 11.

## Test plan

- [x] `go build ./pkg/executor/... ./pkg/runner/...`
- [x] `go vet ./pkg/executor/... ./pkg/runner/...`
- [x] `go test ./pkg/executor/... -count=1 -p 1` (all packages pass)
- [x] New unit tests: `chunkWriter` newline emission, multi-line single write, 4 KB threshold trigger, partial line carry-over, `Flush` remainder, `Bytes()` full output, `Write` length return
- [x] New integration tests: real `/bin/sh -c` invocation through `ExecWithStreamingHook` verifies chunk ordering and concatenation equals the returned combined output
- [x] Shell streaming regression tests: callback forwarded for single command, callback invoked once per sub-command across `&&` / `||` / `;` with a caller-owned monotonic counter (locks in the seq-collision fix), nil callback falls back to legacy path
- [x] `go run ./cmd/alpamon` boots cleanly against the existing config (connection refused only because no local alpacon-server is running, which is unrelated to this change)

## Notes

- This depends on the server-side `/api/events/commands/{id}/chunk/` endpoint being deployed. Under streaming, command stdout/stderr is delivered solely via chunk POSTs—the fin payload no longer carries the full combined output, only short diagnostics (e.g. privilege-demotion errors and the timeout banner). This keeps the executor from retaining unbounded command output in memory.
- Because output is not teed into fin, the chunk endpoint is required for output delivery: if it is unavailable, successful command output is not recovered via fin. The console degrades to diagnostics-only rather than the full pre-streaming output.

### CODE REVIEW

File: pkg/runner/command.go [hunk 1]
```diff
@@ -53,7 +53,8 @@ func (cr *CommandRunner) Run(ctx context.Context) error {
 		if cr.command.ID != "" {
 			finURL := fmt.Sprintf(eventCommandFinURL, cr.command.ID)
 			payload := protocol.NewCommandResponse(exitCode == 0, result, time.Since(start).Seconds(), exitCode)
-			scheduler.Rqueue.Post(finURL, payload, 10, time.Time{})
+			// Priority 11 > 10 so trailing chunks drain before fin.
+			scheduler.Rqueue.Post(finURL, payload, 11, time.Time{})
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/runner/command.go]: Good catch — reworded to "Lower priority than chunks (10) so trailing chunks drain before fin." in 9632e7f.

File: pkg/executor/executor.go [hunk 2]
```diff
@@ -96,11 +158,23 @@ func (e *Executor) Execute(ctx context.Context, opts CommandOptions) (int, strin
...
+		cw := newChunkWriter(chunkCallback)
+		cmd.Stdout = cw
+		cmd.Stderr = cw
+
+		if err := cmd.Start(); err != nil {
+			return cw.Bytes(), err
+		}
+		if pidHook != nil && cmd.Process != nil {
+			pidHook(cmd.Process.Pid)
+		}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go]: Agreed — extracted invokePIDHook helper that wraps the call in a deferred recover, applied to both the streaming and plain Start/Wait paths in 9632e7f. Behavior now matches the docstring.

File: pkg/executor/executor.go [hunk 3]
```diff
@@ -7,13 +7,75 @@ import (
...
+				w.callback(line)
+			} else {
+				w.buf.WriteString(line)
+			}
+		}
+		break
+	}
+
+	return len(p), nil
+}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go:149]: Good catch — fixed in 0dc8fdb. The threshold path now slices the partial line into chunkSizeThreshold-sized pieces and keeps any <4KB tail buffered, so a single very long no-newline line no longer produces an arbitrarily large payload. Also added panic recovery (see the reply on the Flush comment for details).

File: pkg/executor/executor.go [hunk 4]
```diff
@@ -7,13 +7,75 @@ import (
...
+
+func (w *chunkWriter) Flush() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.buf.Len() > 0 {
+		content := w.buf.String()
+		w.buf.Reset()
+		w.callback(content)
+	}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go:168]: Agreed — extracted an emit() helper on chunkWriter that nil-guards the callback and wraps the call in a deferred recover, then routed all three call sites (newline path, threshold path, and Flush) through it. Behavior now matches PIDHook. Done in 0dc8fdb.

File: pkg/executor/executor.go [hunk 5]
```diff
@@ -7,13 +7,92 @@ import (
...
+		// tail stays buffered for the next Write or Flush.
+		for len(line) >= chunkSizeThreshold {
+			w.emit(line[:chunkSizeThreshold])
+			line = line[chunkSizeThreshold:]
+		}
+		if len(line) > 0 {
+			w.buf.WriteString(line)
+		}
+		break
+	}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go]: Resolved in a later commit. The newline path now routes through `emitChunked(line)`, which slices any line into `chunkSizeThreshold`-sized pieces before emitting. A newline-terminated line that exceeds the threshold (e.g. a 100KB JSON line) is therefore split into ~4KB chunks rather than sent as one oversized payload — the "newline boundary OR 4KB, whichever comes first" contract now holds for both terminated and unterminated lines.

File: pkg/executor/chunk_writer_test.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	}
+	if chunks[0] != strings.Repeat("x", chunkSizeThreshold) {
+		t.Errorf("emitted chunk should be exactly chunkSizeThreshold bytes")
+	}
+
+	cw.Flush()
+	if len(chunks) != 2 || chunks[1] != strings.Repeat("x", 10) {
+		t.Errorf("flush should emit 10-byte tail, got %v", chunks)
+	}
+}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/chunk_writer_test.go:116]: Added `TestChunkWriter_NewlineLineExceedingThresholdSplits`, which writes a sub-threshold tail followed by a long run ending in `\n` (so the newline lands past the 4KB boundary) and asserts the writer emits an exact `chunkSizeThreshold` chunk first, then the remainder retaining the trailing newline. This locks in the "threshold-first, no oversized newline-terminated payload" behavior.

File: pkg/executor/executor.go [hunk 7]
```diff
@@ -87,48 +174,78 @@ func (e *Executor) Execute(ctx context.Context, opts CommandOptions) (int, strin
...
 
-// runCommand executes cmd and returns its combined stdout/stderr output.
-// When pidHook is non-nil, the command is started with cmd.Start() so the
-// child's pid can be reported before Wait blocks. When pidHook is nil,
-// the simpler cmd.CombinedOutput() path is used unchanged.
-func (e *Executor) runCommand(cmd *exec.Cmd, pidHook func(pid int)) ([]byte, error) {
-	if pidHook == nil {
+// runCommand uses Start/Wait when streaming or pid reporting is needed;
+// otherwise CombinedOutput. The streaming path returns no bytes.
+func (e *Executor) runCommand(cmd *exec.Cmd, pidHook func(pid int), cw *chunkWriter) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go]: Good catch on the contradiction. We deliberately do **not** tee stdout/stderr into a buffer here: an earlier review round flagged unbounded memory retention for large streams (now guarded by `TestChunkWriter_LargeStreamDoesNotRetainBody`), and teeing would reintroduce exactly that. So rather than change the code to match the stale description, we updated the PR description to match the implemented contract: under streaming, output is delivered solely via chunk POSTs, and fin carries only short diagnostics (demotion errors, timeout banner), not the full combined output. The chunk endpoint is therefore required for output delivery. Resolving as a description fix.

File: pkg/executor/executor.go [hunk 8]
```diff
@@ -47,12 +123,23 @@ func (e *Executor) Execute(ctx context.Context, opts CommandOptions) (int, strin
...
 		if err != nil {
 			log.Error().Err(err).Msg("Failed to demote privileges")
-			return 1, err.Error(), err
+			msg := err.Error()
+			if cw != nil {
+				// In-band so streaming UIs don't see an empty terminal then fin.
+				cw.emit("alpamon: " + msg + "\n")
+				return 1, "", err
+			}
+			return 1, msg, err
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go:239]: Fixed in f4bfef7. The streaming demotion-failure path now returns `msg` in `result` (keeping the in-band `cw.emit` for the live terminal), so the fin payload still carries the diagnostic when chunk delivery fails. This matches the streaming timeout path, which already returns its banner in `result`.

File: pkg/runner/command.go [hunk 9]
```diff
@@ -92,14 +94,29 @@ func (cr *CommandRunner) Run(ctx context.Context) error {
...
+			chunkURL := fmt.Sprintf(eventCommandChunkURL, commandID)
+			// Runner owns seq so chunks across shell operators share one series.
+			var seq int
+			chunkCallback = func(content string) {
+				scheduler.Rqueue.Post(chunkURL, &protocol.CommandChunk{
+					Seq:     seq,
+					Content: content,
+				}, 10, time.Time{})
+				seq++
+			}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/runner/command.go:112]: Fixed in 8cb8cd6. `seq` is now captured and advanced (`s := seq; seq++`) before the `scheduler.Rqueue.Post(...)` call, so even if `Post` panics (recovered upstream in `chunkWriter.emit`), `seq` has already moved on and the next chunk cannot reuse it — no server-side `(command, seq)` collision under best-effort delivery.

File: pkg/executor/executor.go [hunk 10]
```diff
@@ -6,13 +6,89 @@ import (
...
+		line, err := w.buf.ReadString('\n')
+		if err == nil {
+			// A buffered tail merged with this Write can exceed the cap.
+			w.emitChunked(line)
+			continue
+		}
+		for len(line) >= chunkSizeThreshold {
+			w.emit(line[:chunkSizeThreshold])
+			line = line[chunkSizeThreshold:]
+		}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go]: Fixed in 2b63900. `emit` now applies `strings.Clone(content)` before invoking the callback, so a chunk sliced from a larger `line` (e.g. a huge newline-free read) no longer shares the full backing array — each queued chunk retains only its own bytes.

File: pkg/executor/executor.go [hunk 11]
```diff
@@ -6,13 +6,89 @@ import (
...
+
+func (w *chunkWriter) emitChunked(content string) {
+	for len(content) > chunkSizeThreshold {
+		w.emit(content[:chunkSizeThreshold])
+		content = content[chunkSizeThreshold:]
+	}
+	if len(content) > 0 {
+		w.emit(content)
+	}
+}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/executor.go]: Fixed in 2b63900. Since every chunk path (newline split, threshold split, and `Flush` remainder) funnels through `emit`, cloning there covers `emitChunked` too — the substrings handed to the callback no longer pin the source `content` allocation while queued.

**Review by eunyoung14 — COMMENTED** [review]
## Review — output delivery reliability & resource usage

The `chunkWriter` mechanics are solid: memory stays bounded (the `strings.Clone` avoids pinning a large backing array, and `TestChunkWriter_LargeStreamDoesNotRetainBody` locks that in), the caller-owned `seq` counter keeps a single monotonic series across shell operators (`&&`/`||`/`;`), and the callback panic recovery is in place. A few things to consider before merge, all scoped to alpamon itself.

### 🟠 One POST per line — unbatched, on a shared bounded queue

`chunkWriter` emits a chunk on every newline; the 4 KB threshold only *caps* a chunk, it does not *batch* small lines. So a high-volume command (`yes`, `seq 1 10000000`, `cat large.log`) produces roughly **one HTTP POST per output line**.

Those POSTs go through the shared `scheduler.Rqueue` (`MaxQueueSize = 36000`, drained by 4 reporters). When the queue is full, `request()` logs-and-drops the entry with no retry. Consequences:

- A single chatty command can saturate the queue that is shared with **all other agent telemetry** (acks, fins, command results, events) → unrelated reports get dropped agent-wide.
- `chunkWriter.Write` always returns immediately and `Post` never blocks, so the child process gets **no backpressure** — alpamon converts the firehose into unbounded queue churn.

Suggestion: add a batching window to `chunkWriter` (coalesce up to N lines / 4 KB / ~50–100 ms into one chunk), and optionally apply real backpressure when the queue is near capacity so the child is throttled rather than silently dropped.

### 🟠 No retained copy when chunk delivery fails

Under streaming, a successful command's fin payload carries an empty `result` (`runCommand` returns `nil` bytes on the `cw != nil` path) and the output is delivered **solely** via best-effort chunk POSTs. Nothing retains the output, so a dropped or retry-exhausted chunk POST (see the queue concern above) means that output is **permanently unrecoverable** — there is no fallback.

Suggestion: tee a bounded buffer (cap, e.g. 1 MiB, truncate-middle) into the fin payload so a complete-up-to-cap copy is delivered reliably in a single request, independent of per-chunk delivery. This keeps the memory bound this PR introduced while removing the "all-or-nothing per chunk" fragility.

### Notes (non-blocking)

- `command.go:56-57` honestly documents that the priority bump (10→11) is a best-effort hint, not an ordering guarantee — with 4 concurrent reporters a chunk can still land after fin. That's fine as long as the receiver tolerates post-fin, out-of-order arrival keyed by `(command_id, seq)`.
- stdout and stderr are merged into one stream (parity with the old `CombinedOutput`), so the receiver can't distinguish them — worth noting as a known characteristic.
- Test coverage for the streaming path is good; a test exercising queue-pressure / drop-on-full behavior would be a nice addition given the concern above.

**Comment by jisung-02** [comment]
## Addressed review feedback

Thanks for the review — summary of the changes made since then (all on this branch).

### 🟠 One POST per line / no backpressure — fixed (`29d0648`)
- `chunkWriter` now coalesces output and emits on the 4 KB threshold **or** a ~100 ms flush tick, whichever comes first, so a chatty command no longer produces roughly one POST per output line.
- Added `RequestQueue.PostChunk`: while the shared queue is at/above a high-water mark (80% of `MaxQueueSize`), chunk POSTs apply backpressure — throttling the producing command through its blocked stdout pipe — instead of flooding the queue and starving other telemetry. It falls back to dropping past a max wait or on context cancellation so a command can't stall indefinitely. Only chunk traffic uses this path; acks/fins/events keep their non-blocking semantics.

### 🟠 No retained copy when chunk delivery fails — fixed (`8d3a5e5`, `ff2c009`)
- Reintroduced a fin audit copy, but **bounded**: a 1 MiB truncate-middle `capBuffer` is teed alongside the chunk stream, keeping the command's start and end and dropping the middle. Retained memory stays bounded regardless of stream size (locked by `TestChunkWriter_StreamsAllWithBoundedCapture` + `capBuffer` tests), so this does not reintroduce the earlier unbounded-retention concern.
- `runCommand`'s streaming path returns this capped copy; the timeout path carries it too (banner appended only when there is output).
- Extended the same guarantee to the `allow_sh=false` operator path (`&&`/`||`/`;`), which was still dropping per-segment output under streaming. Segments are now accumulated and the total re-capped via a shared `utils.TruncateMiddle` / `utils.AuditOutputCap`, so both execution paths deliver an identical ≤1 MiB audit copy.

Net effect: chunks are the live, best-effort channel; the fin payload is the reliable, bounded audit copy.

### Notes from the review
- **Queue-pressure / drop-on-full coverage**: added (`PostChunk` blocks-until-space, drops-after-max-wait, drops-on-cancel), plus a regression test that the operator path's fin result carries the accumulated output.
- **Priority bump is best-effort** and **stdout/stderr are merged into one stream**: left as-is — both are intentional, documented characteristics; the receiver reassembles by `(command_id, seq)`.

### Cleanup
- `4f35b3f` removes a now-unreachable chunk-splitting loop in `flush` — batching moved threshold splitting into `Write`, so `flush` always sees a sub-threshold remainder.

All affected packages build and tests pass, including `-race` on the streaming and queue paths.

**Review by eunyoung14 — DISMISSED** [review]
## Review — Approve ✅

Careful, well-tested implementation. `go build` / `go vet` / the changed-package tests all **pass under `-race`**, gofmt clean. The cross-repo contract with alpacon-server #2150 (seq · chunk size · backpressure · capped-tee · ShellTimeout · error handling · Go conventions) is honored on **all 7 axes**.

**Highlights**
- Panic-safe seq (advanced before `Post`); stdout+stderr tee into a single `chunkWriter` serialized by `w.mu` → no race (confirmed under `-race`).
- Capped-tee: both the `fin` `result` and the capture buffer are bounded to 1 MiB (`TruncateMiddle`), so the audit copy survives even when chunks are dropped/purged.
- Backpressure **drops** (with a `warn`) and reserves headroom for `fin`/ack rather than blocking or growing unbounded — the right call for an audited agent; command completion is never wedged.
- Demotion error is teed both in-band and into `fin`.

### 🟡 SHOULD-FIX (non-blocking, follow-up is fine)
**UTF-8 boundary split** — `pkg/executor/executor.go:124` (`buf.Next(chunkSizeThreshold)`)
- Chunks are cut on a 4 KiB **byte** boundary, so a multi-byte rune straddling it is split; Go's `json.Marshal` then replaces the invalid bytes with `U+FFFD`, which is **unrecoverable on reassembly**.
- ASCII is unaffected, but non-ASCII output (CJK / emoji / TUI progress bars) is corrupted at every ~4 KiB boundary in the live stream. (The 1 MiB `fin` audit copy is durable and only hits its two head/tail boundaries.)
- Fix: back the cut off to the last complete rune before emit — see the inline comment for a snippet. Overhead is negligible (a constant-time ≤3-byte check per chunk). Fine as a follow-up since the audit copy is the durable record.

### ⚪ NIT (optional)
- The operator path holds N×1 MiB transiently before the final truncate (pathological `a;a;a;...`); segment count is small in practice — optional incremental truncate.
- The dropped-chunk `warn` (`queue.go`) could carry a structured `command_id` for greppable gap incidents (the URL embeds the id, so it's recoverable — optional).

Overall solid — contract, safety, and tests all pass. Approving; please pick up the UTF-8 boundary as a follow-up. 👍

File: pkg/executor/executor.go [hunk 12]
```diff
@@ -6,13 +6,163 @@ import (
...
+
+func (w *chunkWriter) Write(p []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	w.capture.write(p)
+	w.buf.Write(p)
+	// Emit full-threshold chunks only; the sub-threshold tail waits for the flush tick.
+	for w.buf.Len() >= chunkSizeThreshold {
+		w.emit(string(w.buf.Next(chunkSizeThreshold)))
```
Comment (eunyoung14) [inline pkg/executor/executor.go]: `buf.Next(chunkSizeThreshold)` cuts on a **byte** boundary, so a multi-byte UTF-8 rune straddling the 4 KiB mark is split across two chunks. Go's `json.Marshal` then replaces the invalid UTF-8 with `U+FFFD`, so the split bytes are corrupted **on the wire** and can't be reassembled by the consumer — streamed CJK / emoji / box-drawing (TUI progress bar) output is mangled at every ~4 KiB boundary. ASCII is unaffected, and the 1 MiB `fin` audit copy only hits its two head/tail boundaries, so this is live-stream-only.

Suggested fix — back the cut off to the last complete rune and let the partial rune ride the next chunk:

```go
b := w.buf.Bytes()
end := chunkSizeThreshold
// Don't split a multi-byte rune at the cut: if the next byte is a UTF-8
// continuation byte we're mid-rune, so back off to the rune start (≤3 bytes;
// runes are ≤4 bytes). The remainder is carried into the next chunk.
for end > 0 && !utf8.RuneStart(b[end]) {
    end--
}
w.emit(string(w.buf.Next(end)))
```

**Overhead is negligible** — this inspects at most the last 3 bytes per emitted chunk (not a per-byte scan of the 4 KiB), so throughput is unaffected. Non-blocking; fine as a follow-up.

Comment (jisung-02) [reply pkg/executor/executor.go]: Fixed in 030f1c9.

Added runeSafeCut to back the cut off to a rune boundary at the 4KB mark; the partial trailing rune rides the next chunk. Applies to the streaming path only; the 1 MiB fin audit copy is unchanged.

One deviation from the suggested snippet: I used DecodeLastRune on the slice being emitted rather than RuneStart(b[end]). Indexing b[chunkSizeThreshold] panics (index out of range) when buf is exactly at the threshold, whereas DecodeLastRune(b[:end]) is bounds-safe and also covers that exact-boundary case. Backoff still inspects only the last <4 bytes, so overhead is negligible as you noted.

Added a regression test (TestChunkWriter_DoesNotSplitRuneAtThreshold) asserting per-chunk UTF-8 validity + lossless reassembly; verified it fails on the old code.

**Review by eunyoung14 — APPROVED** [review]

File: pkg/executor/executor.go (hunk 12 above)
Comment (eunyoung14) [reply pkg/executor/executor.go]: Fixed in 030f1c9 — thanks. Your `runeSafeCut` (DecodeLastRune on the emitted slice) is actually safer than my `RuneStart(b[end])` snippet, which would panic with index-out-of-range when the buffer is exactly at the threshold. Nice catch, and good call adding the 가 + utf8.ValidString test. Resolved.

**Review by eunyoung14 — APPROVED** [review]
## Review — Approve ✅ (re-review at `8f712fc6`)

Re-reviewed after the branch moved. The one SHOULD-FIX I raised — splitting a UTF-8 rune at the 4 KiB chunk boundary — is **resolved in `030f1c9`** (`runeSafeCut`), with a `'가'` + `utf8.ValidString` regression test. Your `DecodeLastRune`-on-the-emitted-slice approach is actually safer than the `RuneStart(b[end])` snippet I suggested, which would panic (index out of range) when the buffer sits exactly at the threshold — nice catch.

Cross-repo contract (with alpacon-server #2150 / alpacon-cli #198) is honored on every axis: seq (panic-safe, single serialized series), ≤4 KiB chunks, backpressure that drops + reserves headroom (never wedges the command), 1 MiB capped-tee audit copy, ShellTimeout, non-blocking error handling. Build / vet / changed-package tests pass under `-race`.

Remaining items are optional NITs only (operator path holds N×1 MiB transiently before the final truncate; the dropped-chunk `warn` could carry a structured `command_id`). LGTM 👍
